kernel/proc/scheduler.md

```diff
@@ -0,0 +1,30 @@
+# Kernel - Processes and threads - Scheduler
+The operating system scheduler controls threads execution with a predetermined policy. It is a part of Phoenix-RTOS having the most significant influence on the performance and responsiveness of the whole system.
+
+## Scheduling policy
+The scheduling algorithm is defined in the `threads_schedule` function, which is invoked by a timer hardware interrupt. Each function's invocation changes the context of the currently executed thread to the next one, available in a thread's ready queue. The scheduling policy determined which thread shall be chosen for the next execution. In Phoenix-RTOS, there is used a priority preempted round-robin algorithm. There are eight priority levels, the smallest value holds the highest priority.
+
+
+The thread management unit contains an eight-element array, where each of them holds a pointer to the list of threads of the same priority. A scheduling algorithm is defined as follows:
+
+1. The spinlock is set before any operations on common thread management data.
+2. The unit holds information about the running threads of each processor's core. The current thread's context for the interrupted core is saved and added to the end of the priority list.
+3. The next available thread with the highest priority is selected to be run next and is removed from the ready thread's list. If a selected thread is a ghost, it is added to the ghosts list and the reaper thread is waked up.
+4. For the selected thread, the following actions are performed:
+    * a pointer to the current thread is changed to the selected one,
+    * a kernel stack is changed to the stack of a new thread,
+    * a memory map is changed to the map associated with the thread's process,
+    * signal handlers are performed,
+    * performance data is saved in a perf unit,
+5. The cpu usage is updated for the current and selected thread.
+6. At the end of the modification of the common thread management data, the spinlock is cleared.
```

Comment:
There is no information about context switch - how and where it's done?

Pointer to the stack in context of current thread is updated with a pointer to the stack of selected thread. When we exit scheduler context restore is performed - thanks to this change in saved SP we restore selected thread instead.